internal/manifest/oci_index.go

```diff
@@ -93,19 +136,38 @@ func (index *OCI1IndexPublic) ChooseInstance(ctx *types.SystemContext) (digest.D
 				Variant:      d.Platform.Variant,
 			}
 			if platform.MatchesPlatform(imagePlatform, wantedPlatform) {
-				return d.Digest, nil
+				if bestMatchedInstance.digest == "" || instanceCandidate.isPreferredOver(bestMatchedInstance, didPreferGzip) {
```

Comment:
This is now unnecessarily called `len(manifests)*len(wantedPlatforms)` times.
- If `d.Platform == nil`, we don’t need to iterate through `wantedPlatforms` at all
- Otherwise, just search for the first match, and use that to set `instanceCandidate.platformIndex`; call `isPreferredOver` once per manifest.

Sure, the current implementation works and is not asymptotically any more costly.

Compare https://github.com/containers/image/pull/1789#discussion_r1113453162 suggesting that the inner loop could be a helper function; I don’t care right now about actually factoring that function out.

My primary goal is that the structure should cleanly separate the `platformIndex` determination from the rest of the logic, so that we can, in the future, replace `wantedPlatforms` with something that allows O(1) lookup.

Reply:
Yes, that commit gets the core point right.